### ISSUE TITLE
TWRP inappropriately locks the eMMC while installing zips

### ISSUE THREAD

**Issue by Lanchon** [TeamWin/Team-Win-Recovery-Project#664]
operations on the eMMC block device cannot be performed in zip packages because TWRP locks the eMMC while installing zips.

for a real world example of the impact of this bug, take a look at my [repartitioning tool for android](https://github.com/Lanchon/REPIT). it is unable to repartition when normally flashed on TWRP. as a workaround, it detects the condition, copies itself to /tmp, and asks you to reflash from there. but this is very awkward.

when flashing a zip, TWRP has to extract 'update-binary' to /tmp and run it. but when the zip is anywhere but in /tmp, TWRP inappropriately holds some resource in use, locking the eMMC. could it be the current directory of some process? maybe the install/extract process itself? maybe the file picker? whatever it is, it causes this bug.

i prepared a zip package for you to test. it unmounts all emmc partitions and forces the kernel to reread the partition table. it works in /tmp and when sideloaded. it fails when flashed from the sdcard. here is the file: [refresh-pt.zip](https://github.com/TeamWin/Team-Win-Recovery-Project/files/233145/refresh-pt.zip)

this script requires the emmc to be /dev/block/mmcblk0, which i think is most common. but it was created with this line, so you can customize it to your test device:

```
echo 'grep -ow "^/dev/block/mmcblk[^ ]*" /proc/mounts | sort -u | xargs -r -n1 umount; blockdev --rereadpt /dev/block/mmcblk0' | flashize - refresh-pt.zip
```

please let me know if there's anything i can help you with.
thank you!

**Comment by that1** [comment]
Installing a zip requires that the partition containing the zip file is kept mounted so that the update-binary can extract files from the zip. This is by design (of the Android recovery).

**Comment by Lanchon** [comment]
that is true, but that is not the issue. the issue is that -besides being mounted- the partition is also unmountable!

some part of twrp keeps something in the partition needlessly in use during flashing (maybe some process' current directory?), and thus unmounting fails.

if you take a look at the code i posted above, you'll see that it unmounts all mounted partitions before acting on the device itself. try it and you'll see that the unmounts fail.

---

PS. there is even an extra complication: if you flash the file from the external sdcard, you'll see that the zip is also unable to reread the partition table of the eMMC, even though the eMMC and the external sdcard are different devices and the mounted sdcard shouldn't affect the eMMC.

**Comment by Lanchon** [comment]
i would like to investigate why the flashed-from file system can't be unmounted, but i don't know how. do you know of a way i could ask the kernel what is being locked and/or by whom?

**Comment by Dees-Troy** [comment]
The original AOSP design for flashing zips included copying the zip from the file system into /tmp (into RAM). We found this design to be problematic because some zips are larger than the amount of free RAM on some devices. It should be theoretically possible to accomplish what you are hoping, but you will probably have to modify or completely replace the updater binary. TWRP mmaps the zip file, opens the zip and copies the updater binary into /tmp, closes the zip, and then executes the updater binary, passing along certain parameters to the updater binary including the path to the zip:

https://github.com/TeamWin/Team-Win-Recovery-Project/blob/android-6.0/twinstall.cpp#L202

The updater binary then mmap's the zip file and opens the zip file as a zip:

https://github.com/TeamWin/Team-Win-Recovery-Project/blob/android-6.0/updater/updater.c#L74

The mmap in TWRP may be locking the file system, so we could munmap it before executing the updater binary, however, it's more likely that the updater binary is to blame because the updater binary usually keeps the zip file open for the duration of executing the update script:

https://github.com/TeamWin/Team-Win-Recovery-Project/blob/android-6.0/updater/updater.c#L191

While these issues do exist within TWRP's updater binary source code, these same issues also exist in any other AOSP-based recovery and thus aren't exactly a "TWRP issue".

https://android.googlesource.com/platform/bootable/recovery/+/android-6.0.1_r43/updater/updater.c#163
https://github.com/CyanogenMod/android_bootable_recovery/blob/cm-13.0/updater/updater.c#L163

Further, modifying the updater binary source code to support this specific use-case is outside of the scope of TWRP.

**Comment by Lanchon** [comment]
hi D,

thank you very much for the detailed and thoughtful reply. unfortunately you partially went down the wrong road here. **i HAVE completely replaced the update binary.** if you look at my first post here, you'll see that 'refresh-pt.zip' does not contain AOSP's update binary.

the 'refresh-pt.zip' file was created by the one liner quoted in that post using [flashize](https://github.com/Lanchon/Flashize). (the flashize runtime accepts the arguments you mention passed by TWRP and destined to the updater binary.)

> The mmap in TWRP may be locking the file system, so we could munmap it before executing the updater binary

i haven't looked at TWRP's code, but i'm looking at it now...

i see a bug here, this line should go, double close:
https://github.com/TeamWin/Team-Win-Recovery-Project/blob/android-6.0/twinstall.cpp#L163

yes, definitely keeping the mmap open during the install is the problem:
https://github.com/TeamWin/Team-Win-Recovery-Project/blob/android-6.0/twinstall.cpp#L331-L338
(in theory this doesn't rule out other locks, but very probably fixing this will fix the issue.)

the solution is to refactor `Run_Update_Binary` into `Extract_Update_Binary` and `Run_Update_Binary` and (conditionally on Extract_\* success) invoke Run_\* after the `sysReleaseMap` call.

of course i could do a pull request with these simple changes (double close and locking) but i hope you don't mind that i rather not sign any copyright session agreements.

**Comment by Lanchon** [comment]
BTW, if you want to go ahead with this fix, remember that you can easily test the change by flashing the zip in the first post. the only requirement is that `/dev/block/mmcblk0` must be the phone's eMMC.

(if your test device doesn't comply, just tell me the name of your eMMC device node and i'll post a zip for you.)
